src/coprocessor/cache.rs

```diff
@@ -0,0 +1,487 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::option::Option;
+use std::sync::Mutex;
+use linked_hash_map::LinkedHashMap;
+use super::metrics::*;
+
+type DistSQLCacheKey = String;
+
+const DISTSQL_CACHE_ENTRY_ADDITION_SIZE: usize = 40;
+
+pub struct DistSQLCacheEntry {
+    region_id: u64,
+    version: u64,
+    result: Vec<u8>,
+    size: usize,
+}
+
+impl DistSQLCacheEntry {
+    pub fn new(region_id: u64, version: u64, key_size: usize, r: Vec<u8>) -> DistSQLCacheEntry {
+        let size = r.len() + DISTSQL_CACHE_ENTRY_ADDITION_SIZE + 2 * key_size;
+        DistSQLCacheEntry {
+            region_id: region_id,
+            version: version,
+            result: r,
+            size: size,
+        }
+    }
+
+    pub fn update(&mut self, region_id: u64, version: u64, key_size: usize, res: Vec<u8>) -> usize {
+        let old_size = self.size;
+        self.size = res.len() + (key_size * 2) + DISTSQL_CACHE_ENTRY_ADDITION_SIZE;
+        self.version = version;
+        self.result = res;
+        self.region_id = region_id;
+        self.size - old_size
+    }
+}
+
+// RegionDistSQLCacheEntry track Region's cache items key.
+// So we can quickly get a cache entry list by given a region id.
+// That can make evict_region operation more faster than scan whole
+// cache items and get the remove item list.
+pub struct RegionDistSQLCacheEntry {
+    version: u64,
+    enable: bool,
```

Comment:
It seems we could drop the `RegionDistSQLCacheEntry ` when enable=false?

Reply:
No, enable=false means this region's status is between pre_apply and post_apply, the disk writing is not finished. So we disable writing region's cache to prevent cache wrong data.